<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>dmr_analysis_block &mdash; dmr_analysis 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dmr_combine_multChrs4rank" href="dmr_combine_multChrs4rank.html" />
    <link rel="prev" title="installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            dmr_analysis
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="README.html">README</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">dmr_analysis_block</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#required-parameters">Required Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optional-parameters">Optional Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dmr_combine_multChrs4rank.html">dmr_combine_multChrs4rank</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_selected4plot.html">dmr_selected4plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_map2genome.html">dmr_map2genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_map2chromSegment.html">dmr_map2chromSegment</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_cal2genome_percent.html">dmr_cal2genome_percent</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_cal2chromSegment_percent.html">dmr_cal2chromSegment_percent</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_percent2plot.html">dmr_percent2plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_combine2geneAnnot.html">dmr_combine2geneAnnot</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_exportData.html">dmr_exportData</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_gene_annotation.html">dmr_gene_annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="check_accuracy4cluster.html">check_accuracy4cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo1.html">demo1</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo2.html">demo2</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo3.html">demo3</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dmr_analysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">dmr_analysis_block</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/dmr_analysis_block.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dmr-analysis-block">
<h1>dmr_analysis_block<a class="headerlink" href="#dmr-analysis-block" title="Permalink to this heading"></a></h1>
<p>Significant test of TF binding affinity changes between the foreground and the background calculations</p>
<section id="required-parameters">
<h2>Required Parameters<a class="headerlink" href="#required-parameters" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-in</span> <span class="pre">IN_FILE_FOLDER,</span> <span class="pre">--in_file_folder</span> <span class="pre">IN_FILE_FOLDER</span></code>: Input file folder for DNA methylation data such as WGBS. In this folder, all samples are provided in each chromosome folder with BED format where sample name is indicated by file name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-chr</span> <span class="pre">CHROMOSOME,</span> <span class="pre">--chromosome</span> <span class="pre">CHROMOSOME</span></code>: Select a chromosome for running dmr_analysis_block. For example, a chromosome (e.g., chrY) file folder under –in_file_folder that contains methylation data of samples in a chromosome (e.g., chrY).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-gkey</span> <span class="pre">GROUP_KEY,</span> <span class="pre">--group_key</span> <span class="pre">GROUP_KEY</span></code>: Group key name, all bed files name contain this group_key will be combined together for dmr_analysis_block. In other words, only bed file name with this group_key will be selected in analysis. Usually, it is the same as the file folder (or chromosome name) in –chromosome.</p></li>
</ul>
</section>
<section id="optional-parameters">
<h2>Optional Parameters<a class="headerlink" href="#optional-parameters" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-out</span> <span class="pre">,</span> <span class="pre">--out_file_folder</span></code>: Output file folder that stores all results computed by dmr_analysis_block. The default is “out/”.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-ncol</span> <span class="pre">,</span> <span class="pre">--need_column</span></code>: Select a column name from the bed file that will be used by dmr_analysis_block. The default is “Methylation”.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-wtStr</span> <span class="pre">,</span> <span class="pre">--wildType_fileString</span></code>: Use the first few character string of a file name to indicate it is a normal/wild type sample, or to label this file/sample as wild type/normal condition. The default is “gcb”.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-dstart</span> <span class="pre">,</span> <span class="pre">--data_start_position</span></code>: Data start position, a start column position for input dataframe after methylation levels of multiple samples are combined into one file. The default is 3 for chrY demo.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-dend</span> <span class="pre">,</span> <span class="pre">--data_end_position</span></code>: Data end position, an end column position for input dataframe after methylation levels of multiple samples are combined into one file. The default is 15 for chrY demo.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-mxL</span> <span class="pre">,</span> <span class="pre">--maximum_adjacency_length</span></code>: Maximum length of adjacency CpG sites allowed in a methylation region (MR). The default is 250.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-minS</span> <span class="pre">,</span> <span class="pre">--minimum_block_size</span></code>: Minimum number of CpG sites requested in a block/methylation region (MR). The default is 5.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-nDP</span> <span class="pre">,</span> <span class="pre">--number_of_data_points</span></code>: The number of data points (or rows from a dataframe) that will be considered in analysis. The default is 0, which means all data points (or the combined dataframe from all samples and CpG sites) will be used.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-pC</span> <span class="pre">,</span> <span class="pre">--P_cutoff</span></code>: P cutoff value for T-test or other statistical significance test used in the analysis. The default is 0.05.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-aC</span> <span class="pre">,</span> <span class="pre">--accuracy_cutoff</span></code>: Clustering accuracy cutoff value for binary classification (e.g., DMR or not DMR; tumor or normal samples) of 2D t-SNE map. The default is &gt; 0.5.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-mPer</span> <span class="pre">,</span> <span class="pre">--minimum_percentage_changes</span></code>: Minimum percentage of data points in a MR that passed a predefined filtering condition such as the methylation changes greater than a predefined cutoff value. The default is 0.0001 (0.1 percentage).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-perC</span> <span class="pre">,</span> <span class="pre">--percentage_cutoff</span></code>: A comma-separated string list for predefined cutoff values of percentage of methylation changes between the two groups (e.g., 0.07,0.15,0.2) for low, median, high changes, respectively. The default parameter is 0.07,0.15,0.2.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-lmhC</span> <span class="pre">,</span> <span class="pre">--low_median_high_cutoff</span></code>: Use low, median, or high percentage changes (from –percentage_cutoff) to predict high confidence DMRs. Use 0, 1, or 2 to represent low, median, or high percentage cutoff for the prediction. The default is 2 (high percentage changes in –percentage_cutoff).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-numPro</span> <span class="pre">,</span> <span class="pre">--number_of_processes</span></code>: Number of parallel processes that will be used in calculation. The default is 1.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-isSmd</span> <span class="pre">,</span> <span class="pre">--is_smoothed_data</span></code>: Use 0, 1, or 2 to represent raw data, interpolated data, or smoothed data used in predicting DMRs when comparing the two groups. The default is 2 (use smoothed data points to predict DMRs).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-isMT</span> <span class="pre">,</span> <span class="pre">--is_moderate_ttest</span></code>: Use 0 for standard T-test, 1 for moderate T-test, and 2 for KS-test for evaluating the significance of differential methylation between the two groups across all CpG sites in a MR. The default is 0 (uses a T-test).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-isExt</span> <span class="pre">,</span> <span class="pre">--is_export_data</span></code>: Use 0 to not export data or 1 to export data during the analysis. Both raw and smoothed methylation levels of all DMRs will be exported. The default is 0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-dotOrUnderscore</span> <span class="pre">,</span> <span class="pre">--column_splitBy_dotOrUnderscore</span></code>: Use 0 to split column labels using a dot (.) or 1 to split column labels using an underscore (_). The default is 0 (using a dot to split column labels).</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="dmr_combine_multChrs4rank.html" class="btn btn-neutral float-right" title="dmr_combine_multChrs4rank" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Omer &amp; Junbai.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>