<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>dmr_selected4plot &mdash; dmr_analysis 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dmr_map2genome" href="dmr_map2genome.html" />
    <link rel="prev" title="dmr_combine_multChrs4rank" href="dmr_combine_multChrs4rank.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            dmr_analysis
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="README.html">README</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_analysis_block.html">dmr_analysis_block</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_combine_multChrs4rank.html">dmr_combine_multChrs4rank</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">dmr_selected4plot</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#required">Required</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optional-parameters">Optional Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dmr_map2genome.html">dmr_map2genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_map2chromSegment.html">dmr_map2chromSegment</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_cal2genome_percent.html">dmr_cal2genome_percent</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_cal2chromSegment_percent.html">dmr_cal2chromSegment_percent</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_percent2plot.html">dmr_percent2plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_combine2geneAnnot.html">dmr_combine2geneAnnot</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_exportData.html">dmr_exportData</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmr_gene_annotation.html">dmr_gene_annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="check_accuracy4cluster.html">check_accuracy4cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo1.html">demo1</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo2.html">demo2</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo3.html">demo3</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dmr_analysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">dmr_selected4plot</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/dmr_selected4plot.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dmr-selected4plot">
<h1>dmr_selected4plot<a class="headerlink" href="#dmr-selected4plot" title="Permalink to this heading"></a></h1>
<p>Plot figures or export methylation data for selected DMR/MRs.</p>
<section id="required">
<h2>Required<a class="headerlink" href="#required" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-inDMR</span> <span class="pre">IN_DMR_FILE</span></code>: A file generated after performing ‘dmr_analysis_block’ and ‘dmr_combine_multChrs4rank,’ which contains DMR information and logReg probability values for each MR, in addition to the main feature scores exported from dmr_analysis_block. This file is usually exported for each chromosome under the ‘plots’ folder in the output folder of dmr_analysis_block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inData</span> <span class="pre">IN_DATA_FILE</span></code>: A file containing raw DNA methylation data of all MRs in a chromosome, exported after running ‘dmr_analysis_block.’ This file is stored in each chromosome’s output folder. For example, a file named ‘chrY_MR_data4maxBlockDistance_250_minBlockSize_5_data.txt.gz’ is stored in the ‘chrY’ output folder after running ‘dmr_analysis_block.’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inFolder</span> <span class="pre">IN_DATA_FOLDER</span></code>: Path of a file folder that contains all the files mentioned in <code class="docutils literal notranslate"><span class="pre">-inDMR</span></code> and <code class="docutils literal notranslate"><span class="pre">-inData</span></code>. For example, ‘out/chrY’ is the path that stores DMR/MR raw data for the ‘chrY’ chromosome after running dmr_analysis_block and dmr_combine_multChrs4rank.</p></li>
</ul>
</section>
<section id="optional-parameters">
<h2>Optional Parameters<a class="headerlink" href="#optional-parameters" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-inDMRFolder</span></code>: A subfolder for storing exported MR/DMRs in each chromosome after running dmr_analysis_block. The default is ‘plots/’ under <code class="docutils literal notranslate"><span class="pre">--in_data_folder</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inAcCut</span></code>: A range of clustering accuracy to consider in the program. The default is ‘0.0,1.0,’ which means between 0.0 and 1.0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inDmrSt</span></code>: A string used in <code class="docutils literal notranslate"><span class="pre">--inDMR_file</span></code> that represents the predicted DMR. The default is ‘D,’ representing predicted DMR.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inOtCod</span> <span class="pre">IN_OTHER_CONDITION</span></code>: A string type for selecting DMR from the whole data, such as logReg_proba, min_dmr,max_dmr, top_to_down, top_rank, down_rank, or None. The default is down_rank_10 (10%).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inDstart</span></code>: The position of the data start column in a file inputted from <code class="docutils literal notranslate"><span class="pre">--in_data_file</span></code>. The default is 3 because the first three columns are chromosome positions (chrom, start_post, end_post), followed by data columns (data1, data2, …).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inIsExt</span></code>: Whether to export data (1) or not (0) for selected DMR/MRs. The default is 0, indicating not exporting data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-dotOrUnderscore</span></code>: Use 0 for dot (‘.’) split column labels or 1 for underscore (‘_’) split column labels in the file specified by <code class="docutils literal notranslate"><span class="pre">--in_data_file</span></code>. The default is 0 (dot split column labels).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inIsPlt</span></code> or <code class="docutils literal notranslate"><span class="pre">--is_plot</span></code>: Whether to plot the figure (1) or not (0) for selected DMR/MRs. The default is 0, indicating not plotting the figure.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inIsPas</span></code> or <code class="docutils literal notranslate"><span class="pre">--is_pause</span></code>: Whether to pause (1) or continue (0) in the loop. The default is 0, indicating continuing in the loop for all input DMR/MRs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inLogNm</span></code> or <code class="docutils literal notranslate"><span class="pre">--in_Logger_number</span></code>: Logger number for recording screen displays. The default is 0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inOutFd</span></code> or <code class="docutils literal notranslate"><span class="pre">--out_folder</span></code>: Path for the output file folder. The default is ‘out/’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-inNdChk</span></code> or <code class="docutils literal notranslate"><span class="pre">--needs_check_mr</span></code>: String ID of selected DMR/MRs that will be checked/plotted/data exported, comma-separated. For example, ‘mr10,mr111.’ The default is not to check.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-dpi</span></code> or <code class="docutils literal notranslate"><span class="pre">--figure_dpi</span></code>: Export figure resolution in DPI. The default dpi is 30.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-format</span></code> or <code class="docutils literal notranslate"><span class="pre">--figure_format</span></code>: File format of the exported figure, such as jpg, svg, pdf, or png. The default is pdf.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-wtStr</span></code> or <code class="docutils literal notranslate"><span class="pre">--wildType_fileString</span></code>: A file name of the wild-type condition file, which should start with specific characters. For example, if a file name starts with ‘gcb_meth1_*’ is a wild-type/control sample, then <code class="docutils literal notranslate"><span class="pre">--wildType_fileString</span></code> is ‘gcb,’ which is the default setting in the program.</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="dmr_combine_multChrs4rank.html" class="btn btn-neutral float-left" title="dmr_combine_multChrs4rank" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="dmr_map2genome.html" class="btn btn-neutral float-right" title="dmr_map2genome" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Omer &amp; Junbai.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>