# Differential Methylated Region Analysis Tool Documentation

dmr_analysis is a software tool for differentially Methylated Regions analysis to rank significant DMRs.



[Home](index.md) | [DMR Block Analysis](dmr_analysis_block.md) | [Combine MultiChr4Rank](dmr_combine_multChrs4rank.md) | [Selected4plot](dmr_selected4plot.md) | [map2genome](dmr_map2genome.md) | [map2chromSegment](dmr_map2chromSegment.md) | [cal2genome_percent](dmr_cal2genome_percent.md) | [cal2chromSegment_percent](dmr_cal2chromSegment_percent.md) | [percent2plot](dmr_percent2plot.md) | [combine2geneAnnot](dmr_combine2geneAnnot.md) | [exportData](dmr_exportData.md) | [gene annotation](dmr_gene_annotation.md)
## map2genome:

Map DMR/MRs to predefined genomic regions such as TSS, TES, 5dist et al.

### Required Parameters:
<ul>
  <li><code> in_cluster_out_path: </code> File path contains all clustered PWMs and uncertain
                        PWMs, such as '../data/out/sep8_run1</li>
<li><code> in_bpb3_ranking_path: </code> File path for bpb3 ranking results, such as ../data/ou
                        t/sep8_run1/test_67snps/foreground/test_seq/rankings/</li>
  <li><code> in_fileString4clustered_pwm: </code> File string or file folder/name used/contained for
                        clustered PWMs such as quality_assessed_out_seed3</li>

    
</ul>

<p><strong>Required:</strong></p>
<ul>
  <li><code>-inSDFile IN_SORTEDDMR_FILE, --in_sortedDMR_file IN_SORTEDDMR_FILE</code> - An exported file from DMR_analysis with BED format, such as exported file from "dmr_combine_multChrs4rank" where columns are chrom, start_position, end_position, DMR_ID, probability values. Pleaes note this input file shall be sorted by chromosome position before it is being intersected with predefined genomic regions such as TSS, gene, or TES et al.</li>
  <li><code>-inGRFile IN_GENEREGION_FILE, --in_geneRegion_file IN_GENEREGION_FILE</code> - A list of gene region files that is generated by hmst_seq_analyszer such as "list_region_files.txt" where file name and path of genomic region files are listed. These predefined genomic region files will be intersected with sorted bed file of MR/DMR, please refer to demo of "run_hmsq.sh" for how to generate these predefined genomic region/annotation files.</li>
  <li><code>-inOFolder IN_OUTFILE_FOLDER, --in_outFile_folder IN_OUTFILE_FOLDER</code> - Path of output file folder for exporting intersected MR/DMR files with each genomic regions.</li>
  <li><code>-inRefFile IN_REFFLAT_FILE, --in_refFlat_file IN_REFFLAT_FILE</code> - Path of input refFlat reference file, such as an export file "hg19.refFlat_clean_sorted.bed" from "hmst_seq_Analyzer"</li>
</ul>

<p><strong>Optional , has default values:</strong></p>
<ul>
  <li><code>-minOlp , --in_minimum_overlap4bedtools</code> - Minimum overlap rate in bedtools intersection, default is 1e-9 or 1bp overlap, which is float number between 0 and 1.</li>
</ul>


[Home](index.md)
